beginner_source/ax_multiobjective_nas_tutorial.py

```diff
@@ -0,0 +1,524 @@
+we may want to maximize model performance (e.g. accuracy), while
+simultaneously minimizing competing metrics like power consumption,
+inference latency, or model size in order to satisfy deployment
+constraints. Often, we may be able to reduce computational requirements
+or latency of predictions substantially by accepting minimally lower
+model performance. Principled methods for exploring such tradeoffs
+efficiently are key enablers of scalable and sustainable AI, and have
+many successful applications at Meta - see for instance our
+`case study <https://research.facebook.com/blog/2021/07/optimizing-model-accuracy-and-latency-using-bayesian-multi-objective-neural-architecture-search/>`__
+on a Natural Language Understanding model.
+
+In our example here, we will tune the widths of two hidden layers,
+the learning rate, the dropout probability, the batch size, and the
+number of training epochs. The goal is to trade off performance
+(accuracy on the validation set) and model size (the number of
+model parameters).
+
+This tutorial makes use of the following PyTorch libraries:
+
+- `PyTorch Lightning <https://github.com/PyTorchLightning/pytorch-lightning>`__ (specifying the model and training loop)
+- `TorchX <https://github.com/pytorch/torchx>`__ (for running training jobs remotely / asynchronously)
+- `BoTorch <https://github.com/pytorch/botorch>`__ (the Bayesian Optimization library powering Ax's algorithms)
+"""
+
+
+######################################################################
+# Defining the TorchX App
+# -----------------------
+#
+# Our goal is to optimize the PyTorch Lightning training job defined
+# in `mnist_train_nas.py <https://github.com/pytorch/tutorials/tree/master/beginner_source/mnist_train_nas.py>`__.
+# To do this using TorchX, we write a helper function that takes in
+# the values of the architcture and hyperparameters of the training
+# job and creates a TorchX AppDef (see the
+# `TorchX docs <https://pytorch.org/torchx/latest/basics.html>`__
+# for more detail) with the appropriate settings.
+#
+
+
+from pathlib import Path
+
+from torchx import specs
+from torchx.version import TORCHX_IMAGE
+
+
+# load in the training code
+with Path.cwd().joinpath("mnist_train_nas.py").open("rt") as f:
+    train_source = f.read()
+
+
+def trainer(
+    log_path: str,
+    hidden_size_1: int,
+    hidden_size_2: int,
+    learning_rate: float,
+    epochs: int,
+    dropout: float,
+    batch_size: int,
+    trial_idx: int = -1,
+) -> specs.AppDef:
+
+    # define the log path so we can pass it to the TorchX AppDef
+    if trial_idx >= 0:
+        log_path = Path(log_path).joinpath(str(trial_idx)).absolute().as_posix()
+
+    return specs.AppDef(
+        name="trainer",
+        roles=[
+            specs.Role(
+                name="trainer",
+                image=TORCHX_IMAGE,
+                entrypoint="python",
+                args=[
+                    "-c",
+                    train_source,
```

Comment:
any reason you're passing the train_source instead of just passing the script path?

might also be good to use the python builtin component to reduce amount of TorchX knowledge required https://pytorch.org/torchx/main/components/utils.html#torchx.components.utils.python

Reply:
> any reason you're passing the train_source instead of just passing the script path?

Hmm good point - this is likely a remnant from the original colab setup where the script code was defined in the notebook. Let me see if I can simplify this. 

> might also be good to use the python builtin component to reduce amount of TorchX knowledge required

good idea

Reply:
> might also be good to use the python builtin component to reduce amount of TorchX knowledge required

Actually, I'm not sure this is easily possible. The current `trainer()` function does some dynamic parsing of the trial index to generate the log path that is then passed to the training script. AFAICT it seem that the python builtin component static and wouldn't allow for this parsing. The other option would of course be to pass only the base log path to the script and add an additional `trial_idx` cmd line arg, but ideally doing that wouldn't be necessary. Is there any way to pass dynamic transformations as args to the builtin python component?

Reply:
Nvm I don't think this question makes much sense now that I think about it, I can just redefine `trainer` in the same way.